File: leetcode-04/solutions/python3/0221.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0

        m = len(matrix)
        n = len(matrix[0])

        dp = [0] * n
        max_ = 0
        prev = 0

        for i in range(m):
            for j in range(n):
                temp = dp[j]
                if i == 0 or j == 0 or matrix[i][j] == '0':
                    dp[j] = ord(matrix[i][j]) - ord('0')
                else:
                    dp[j] = min(dp[j], dp[j - 1], prev) + 1
                max_ = max(max_, dp[j])
                prev = temp

        return max_ * max_

File: leetcode-04/solutions/cpp/0221.cpp
class Solution {
 public:
  int maximalSquare(vector<vector<char>>& matrix) {
    if (matrix.empty()) return 0;

    const int m = matrix.size();
    const int n = matrix[0].size();

    vector<int> dp(n);
    int max = 0;
    int prev = 0;

    for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++) {
        int temp = dp[j];
        dp[j] = (i == 0 || j == 0 || matrix[i][j] == '0')
                    ? matrix[i][j] - '0'
                    : min(dp[j], min(dp[j - 1], prev)) + 1;
        max = std::max(max, dp[j]);
        prev = temp;
      }

    return max * max;
  }
};